tests/MongoDB.Bson.Tests/Serialization/SerializationTests.cs

```diff
@@ -234,5 +248,126 @@ public DerivedWithoutSetter_IntermediateBaseAndWithProtectedConstructor(int? x,
+        {
+            [BsonConstructor]
+            public DerivedImmutableWithMorePropertiesThanInConstructorButWithBsonElementAttributeAndWithBsonConstructorAttribute_AbstractBaseImmutable(int y)
+            {
+                Y = y;
+            }
+
+            [BsonElement]
+            public int Y { get; }
+        }
+
+        public abstract class AbstractBaseImmutableWithConstructor
+        {
+            public AbstractBaseImmutableWithConstructor(int x)
+            {
+                X = x;
+            }
+
+            public int X { get; } = 0; // should be overwritten
+        }
+
+        public class DerivedImmutableWithMorePropertiesThanInConstructorAndWithBsonConstructorAttribute_AbstractBaseImmutableWithConstructor : AbstractBaseImmutableWithConstructor
+        {
+            [BsonConstructor]
+            public DerivedImmutableWithMorePropertiesThanInConstructorAndWithBsonConstructorAttribute_AbstractBaseImmutableWithConstructor(int x, int y) : base(x)
+            {
+                Y = y;
+            }
+
+            public int Y { get; }
+        }
+
+        public class DerivedImmutableWithMorePropertiesThanInConstructorButWithBsonElementAttributeAndWithBsonConstructorAttribute_AbstractBaseImmutableWithConstructor : AbstractBaseImmutableWithConstructor
+        {
+            [BsonConstructor]
+            public DerivedImmutableWithMorePropertiesThanInConstructorButWithBsonElementAttributeAndWithBsonConstructorAttribute_AbstractBaseImmutableWithConstructor(int x, int y) : base(x)
+            {
+                Y = y;
+            }
+
+            [BsonElement]
+            public int Y { get; }
+        }
+
+        public class DerivedImmutableWithMorePropertiesThanInConstructorButWithBsonIgnoreAttribute_AbstractBaseImmutableWithConstructor : AbstractBaseImmutableWithConstructor
+        {
+            public DerivedImmutableWithMorePropertiesThanInConstructorButWithBsonIgnoreAttribute_AbstractBaseImmutableWithConstructor(int x) : base(x)
+            {
+            }
+
+            [BsonIgnore]
+            public int Y { get; } = 2;
+        }
+
+
+        public abstract class AbstractBaseImmutableWithAbstractProperty
+        {
+            [BsonIgnore]
+            public abstract int X { get; }
+        }
+
+        public class DerivedImmutableWithMorePropertiesThanInConstructorButWithBsonIgnoreAttribute_AbstractBaseImmutableWithAbstractProperty : AbstractBaseImmutableWithAbstractProperty
+        {
+            public DerivedImmutableWithMorePropertiesThanInConstructorButWithBsonIgnoreAttribute_AbstractBaseImmutableWithAbstractProperty(int y)
+            {
+                Y = y;
+            }
+
+            [BsonIgnore]
+            public override int X { get; } = 1;
+
+            public int Y { get; } = 0; // should be overwritten
+        }
+
+        public class ClassWithTwoConstructorsWhereTheFirstIfNotFullyMatchedAndTheSecondIsFullyMatched
```

Comment:
I think you meant `...IsNotFullyMatched...` instead of `...IfNotFullyMatched...`